### PR TITLE
types/logger: add Context and related helpers

### PR DESCRIPTION
**Pull request by josharian**
We often need both a log function and a context.
We can do this by adding the log function as a context value.
This commit adds helper glue to make that easy.
It is designed to allow incremental adoption.

Updates tailscale/corp#3138

Signed-off-by: Josh Bleecher Snyder <josh@tailscale.com>

### CODE REVIEW

**Comment by josharian** [comment]
cc @danderson @crawshaw who also have Opinions about logging

File: types/logger/logger.go [hunk 1]
```diff
@@ -18,13 +18,36 @@ import (
...
 
+// A Context is a context.Context that should contain a custom log function, obtainable from FromContext.
+// If no log function is present, FromContext will return log.Printf.
+// To construct a Context, use Add
+type Context context.Context
+
+type logfKey struct{}
+
+// FromContext extracts a log function from ctx.
+func FromContext(ctx Context) Logf {
```
Comment (crawshaw) [inline types/logger/logger.go:38]: so we log a line by writing `logger.FromContext(ctx).Logf("foo")`? seems a teensy bit wordy. no obvious alternatives in the typical go vocab, but maybe something like `logger.Printf(ctx, "foo")`?

Comment (josharian) [reply types/logger/logger.go:38]: For sample usage see https://github.com/tailscale/corp/pull/3151

Comment (crawshaw) [reply types/logger/logger.go:38]: I have had enough different opinions over the years that I'll leave this to someone else.

**Review by bradfitz — APPROVED** [review]

**Pull request merged**